junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/DefaultLauncher.java

```diff
@@ -175,11 +184,12 @@ private Root discoverRoot(LauncherDiscoveryRequest discoveryRequest, String phas
 		}
 	}
 
-	private void execute(Root root, ConfigurationParameters configurationParameters,
-			TestExecutionListener... listeners) {
+	private void execute(UnmodifiableTestPlan unmodifiableTestPlan, TestExecutionListener[] listeners) {
+		Root root = unmodifiableTestPlan.getRoot();
+		ConfigurationParameters configurationParameters = root.getConfigurationParameters();
 		TestExecutionListenerRegistry listenerRegistry = buildListenerRegistryForExecution(listeners);
 		withInterceptedStreams(configurationParameters, listenerRegistry, testExecutionListener -> {
-			TestPlan testPlan = TestPlan.from(root.getEngineDescriptors());
+			TestPlan testPlan = unmodifiableTestPlan.getDelegate();
```

Comment:
Is it necessary to work with the "modifiable" `TestPlan` from this point on, or would it suffice to work with the `UnmodifiableTestPlan` during the execution phase?

Now that I think about, wouldn't it be reasonable to disallow modification of the `TestPlan` by a registered `TestExecutionListener`?

Reply:
Makes sense. Changed in my latest commit.